src/mi.rs

```diff
@@ -119,22 +134,57 @@ fn sc<V>(
     tuple: &[V],
     event_freqs: &[HashMap<V, usize>],
     event_sums: &[usize],
+    num_joint_events: usize,
     joint_freq: usize,
     joint_sum: usize,
+    smoothing: &Smoothing
 ) -> f64
-where
-    V: Eq + Hash,
+    where
+        V: Eq + Hash,
 {
-    let tuple_p = joint_freq as f64 / joint_sum as f64;
+    let tuple_p = match smoothing {
+        &Smoothing::Laplace => (joint_freq + 1) as f64 / (joint_sum + num_joint_events) as f64,
+        _ => joint_freq as f64 / joint_sum as f64
+    };
 
-    let indep_p = tuple
-        .as_ref()
-        .iter()
-        .enumerate()
-        .map(|(idx, v)| {
-            event_freqs[idx][v] as f64 / event_sums[idx] as f64
-        })
-        .fold(1.0, |acc, v| acc * v);
+    let indep_p = match smoothing {
+        &Smoothing::Laplace => {
+            let mut num_events = 0;
+            for i in 0..event_freqs.len() {
+                num_events = num_events + event_freqs[i].len();
+            };
+            tuple
+                .as_ref()
+                .iter()
+                .enumerate()
+                .map(|(idx, v)| {
+                    (event_freqs[idx][v] + 1) as f64 / (event_sums[idx] + num_events) as f64
+                })
+                .fold(1.0, |acc, v| acc * v)
+        },
+        &Smoothing::Alpha => {
+            let mut num_events = 0.0;
+            for i in 0..event_freqs.len() {
+                num_events = num_events + (event_freqs[i].len() as f64).powf(0.75);
+            }
+            tuple
+                .as_ref()
+                .iter()
+                .enumerate()
+                .map(|(idx, v)| {
+                    (event_freqs[idx][v] as f64).powf(0.75) / num_events
```

Comment:
Hmm since the denominator in the CDS formula is sum[count(c)^alpha] rather than sum[count(c)]^alpha I figured it would be wrong to simply get the power of the sum - unless I'm misinterpreting `event_sums[idx]`

Reply:
It seems so, but I think then the above code is also not correct. Since you are taking the number of distinct events per position, whereas you need the smoothed total number of events within a position. So, basically you need `event_sums` where the sum for a position are raised to the given power.

Computing that every time is in a tight look very expensive, because you have to go over all the events the given position. So, here you want to precompute the counts once, which is not a problem, if CDS would be its own data type, it could store such a table.

However, I was looking at the Levy et al. paper, and it does not seem to be applicable to compute-mi. They only raise the context counts to this power. However, we do not have the notion of a focus/context. We could interpret the second variable as context, but I am not sure what this would mean when you have more than two variables (smooth all the contexts?).

We can probably use any method used to smooth MLEs. So, maybe we should add just laplace smoothing for this PR and then explore some other techniques (e.g. Good-Turing?) for future PRs.